s2/incident_edge_tracker.go

```diff
@@ -0,0 +1,173 @@
+type vertexEdge struct {
+	vertex Point
+	edgeID int32
+}
+
+// incidentEdgeTracker is a used for detecting and tracking shape edges that
+// are incident on the same vertex. Edges of multiple shapes may be tracked,
+// but lookup is by shape id and vertex: there is no facility to get all
+// edges of all shapes at a vertex. Edge vertices must compare exactly equal
+// to be considered the same vertex, no tolerance is applied as this isn't
+// intended for e.g.: snapping shapes together, which Builder does better
+// and more robustly.
+//
+// To use, instantiate and then add edges with one or more sequences of calls,
+// where each sequence begins with startShape(), followed by addEdge() calls to
+// add edges for that shape, and ends with finishShape(). Those sequences do
+// not need to visit shapes or edges in order. Then, call incidentEdges() to get
+// the resulting map from incidentEdgeKeys (which are shapeId, vertex pairs) to
+// a set of edgeIds of the shape that are incident to that vertex..
+//
+// This works on a block of edges at a time, meaning that to detect incident
+// edges on a particular vertex, we must have at least three edges incident
+// at that vertex when finishShape() is called. We don't maintain partial
+// information between calls. However, subject to this constraint, a single
+// shape's edges may be defined with multiple sequences of startShape(),
+// addEdge()... , finishShape() calls.
+//
+// The reason for this is simple: most edges don't have more than two incident
+// edges (the incoming and outgoing edge). If we had to maintain information
+// between calls, we'd end up with a map that contains every vertex, to no
+// benefit. Instead, when finishShape() is called, we discard vertices that
+// contain two or fewer incident edges.
+//
+// In principle this isn't a real limitation because generally we process a
+// ShapeIndex cell at a time, and, if a vertex has multiple edges, we'll see
+// all the edges in the same cell as the vertex, and, in general, it's possible
+// to aggregate edges before calling.
+//
+// The tracker maintains incident edges until it's cleared. If you call it with
+// each cell from an ShapeIndex, then at the end you will have all the
+// incident edge information for the whole index. If only a subset is needed,
+// call reset() when you're done.
+type incidentEdgeTracker struct {
+	currentShapeID int32
+
+	nursery []vertexEdge
+
+	// We can and do encounter the same edges multiple times, so we need to
+	// deduplicate edges as they're inserted.
+	edgeMap map[incidentEdgeKey]map[int32]bool
+}
+
+// newIncidentEdgeTracker returns a new tracker.
+func newIncidentEdgeTracker() *incidentEdgeTracker {
+	return &incidentEdgeTracker{
+		currentShapeID: -1,
+		nursery:        []vertexEdge{},
+		edgeMap:        make(map[incidentEdgeKey]map[int32]bool),
+	}
+}
+
+// startShape is used to start adding edges to the edge tracker. After calling,
```

Comment:
This comment is copied from the C++ where the scope is ambiguous.  "After calling" more accurately describes finishShape.  If we stick to this API, let's move the comment there.

Reply:
Done